File: docs/validation/go/app/scenarios/av03_future_version.go
package scenarios

// AV-03: v1.0 reader encountering a v1.1 file fails loudly

import (
	"errors"
	"fmt"

	mc "machine-config-go"
)

func RunAv03FutureVersion(fixturesDir, _ string) (bool, string) {
	fixture := avFixture(fixturesDir, "v1_1_simulated.h5")
	_, err := mc.NewReader(fixture).Parse()
	if err == nil {
		return false, "no error raised for File_Version='1.1'"
	}
	var uv *mc.UnsupportedFileVersionError
	if errors.As(err, &uv) {
		if uv.Version == "1.1" {
			return true, fmt.Sprintf("UnsupportedFileVersionError raised, version=%q", uv.Version)
		}
		return false, fmt.Sprintf("UnsupportedFileVersionError raised but version=%q", uv.Version)
	}
	return false, fmt.Sprintf("wrong error type: %v", err)
}

File: docs/validation/cpp/app/src/scenarios/av03_future_version.cpp
// AV-03: v1.0 reader encountering a v1.1 file fails loudly
#include "av03_future_version.hpp"
#include "machine_config/machine_config.hpp"

namespace av03 {

using namespace machine_config;

scenarios::Result run(const std::filesystem::path& fixturesDir, const std::filesystem::path&) {
    auto fixture = scenarios::avFixture(fixturesDir, "v1_1_simulated.h5");
    try {
        MachineConfigReader{fixture}.parse();
        return {false, "no error raised for File_Version='1.1'"};
    } catch (const std::exception& e) {
        std::string msg = e.what();
        if (msg.find("1.1") != std::string::npos) {
            return {true, std::string("error raised naming version '1.1': ") + msg};
        }
        return {false, std::string("error raised but did not name version '1.1': ") + msg};
    }
}

} // namespace av03
